File: matlab/reservoir.m
clear all
close all
clc

addpath('funcs\')
% we use Lorenz system as an example of machine learning short- and
% long-term prediction tasks
% the dimension of the system
dim = 3;

%% hyperparameters

% read optimized hyperparameters 
% load('./opt_MG_tau17_noise=-8.0.mat')
% 
% n = 200;
% hyperpara_set = opt_result;
% eig_rho = hyperpara_set(1);
% W_in_a = hyperpara_set(2);
% alpha = hyperpara_set(3);
% beta = 10^hyperpara_set(4);
% k = round( hyperpara_set(5)/ n);
% % noise_a = 10^hyperpara_set(6);
% noise_a=10^(noise_a);
% noise_a=0;

% hyperparameters by hand
n = 200;
eig_rho = 0.5;
W_in_a = 0.5;
alpha = 0.5;
beta = 10^-5;
k = 0.5;
noise_a=10^-5;

%% data simulation

% set the washup, training, and testing length
transient_length = 5e3;
washup_length = 5e3;
train_length = 5e4 + washup_length;
testing_length = 10000;
short_prediction_length = 800;

tmax = (transient_length + washup_length + train_length + testing_length + 20);
data_length_all = washup_length + train_length + testing_length + 10;
data = zeros(data_length_all, dim); 

% data simulation
ts_train = func_generate_data_lorenz(tmax + 100);
% discard the transient (chaotic system)
ts_train = ts_train(transient_length+1:end, :); 
% data normalization
ts_train=normalize(ts_train);

data(:, :) = ts_train(1:data_length_all, :);

% plot time series data
figure();
plot_length = 2000;
subplot(3, 1, 1);
plot(data(1:plot_length, 1));
title('lorenz system time series')
ylabel('x')

subplot(3, 1, 2);
plot(data(1:plot_length, 2));
ylabel('y')

subplot(3, 1, 3);
plot(data(1:plot_length, 3));
xlabel('step')
ylabel('z')

% plot attractor
figure();
plot_length_long = 8000;
plot(data(1:plot_length_long, 1), data(1:plot_length_long, 3));
title('lorenz system attractor');

%% reservoir computer configuration
rng('shuffle');
% input matrix
W_in = W_in_a*(2*rand(n,dim)-1);
% hidden network: symmeric, normally distributed, with mean 0 and variance 1.
A = sprandsym(n, k); 

%rescale eig
eig_D=eigs(A,1); %only use the biggest one. Warning about the others is harmless
A=(eig_rho/(abs(eig_D))).*A;
A=full(A);

%% train
disp('training...')
tic;

udata = data;

r_train = zeros(n,train_length - washup_length);
y_train = zeros(dim,train_length - washup_length);
r_end = zeros(n,1);

udata(1:train_length,:) = udata(1:train_length,:)+noise_a*randn(size(udata(1:train_length,:)));

% train_x: input; train_y: target
train_x = zeros(train_length,dim);
train_y = zeros(train_length,dim);
train_x(:,:) = udata(1:train_length,:);
train_y(:,:) = udata(2:train_length+1,:);
train_x = train_x';
train_y = train_y';
% initialize r state with all zeros, also can be chosen as random numebrs
r_all = zeros(n,train_length+1);

% update the r state step by step
for ti=1:train_length
    r_all(:,ti+1) = (1-alpha)*r_all(:,ti) + alpha*tanh(A * r_all(:,ti)+W_in*train_x(:,ti));
end

r_out=r_all(:,washup_length+2:end); 
r_out(2:2:end,:)=r_out(2:2:end,:).^2;
r_end(:)=r_all(:,end); 

r_train(:,:) = r_out;
y_train(:,:) = train_y(1:dim,washup_length+1:end);
% calculate the weights of Wout
Wout = y_train *r_train'*(r_train*r_train'+beta*eye(n))^(-1);

t_train = toc;
disp(['training time: ', num2str(t_train)]);

%% test
disp('testing...')

testing_start = train_length+2;

test_pred_y = zeros(testing_length,dim);
test_real_y = zeros(testing_length,dim);
test_real_y(:,:) = udata(testing_start:(testing_start+testing_length-1),:);

r=r_end;
u = zeros(dim,1);
u(:) = udata(train_length+1,:);
for t_i = 1:testing_length
    r = (1-alpha) * r + alpha * tanh(A*r+W_in*u);
    r_out = r;
    r_out(2:2:end,1) = r_out(2:2:end,1).^2; %even number -> squared
    predict_y = Wout * r_out;
    test_pred_y(t_i,:) = predict_y;
    u(1:dim) = predict_y;
end

%% prediction accuracy and illustration

% we use RMSE (root-mean-square error) to measure the short term prediction precision
error = zeros(short_prediction_length,dim);
error(:,:) = test_pred_y(1:short_prediction_length,:) - test_real_y(1:short_prediction_length,:); 

se_ts = sum( abs(error).^2  ,2);
rmse = sqrt( mean(se_ts) );

% plot short term prediction
figure();
subplot(3, 1, 1);
title('lorenz system short term prediction')
plot(test_real_y(1:short_prediction_length, 1), 'r-', LineWidth=1);
hold on
plot(test_pred_y(1:short_prediction_length, 1), 'b--', LineWidth=1);
ylabel('x')

subplot(3, 1, 2);
plot(data(1:plot_length, 2));
plot(test_real_y(1:short_prediction_length, 2), 'r-', LineWidth=1);
hold on
plot(test_pred_y(1:short_prediction_length, 2), 'b--', LineWidth=1);
ylabel('y')

subplot(3, 1, 3);
plot(data(1:plot_length, 3));
plot(test_real_y(1:short_prediction_length, 3), 'r-', LineWidth=1);
hold on
plot(test_pred_y(1:short_prediction_length, 3), 'b--', LineWidth=1);
xlabel('step')
ylabel('z')
legend('real', 'prediction')

% we use DV (deviation value) to measure the long term prediction precision
% define the boundary of the lattice
max_value = 3;
min_value = -3;

xlim_set = [min_value, max_value];
ylim_set = [min_value, max_value];

lower_ = xlim_set(1);
dv_dt = 0.05;
matrix_size = ceil((xlim_set(2)-xlim_set(1))/dv_dt);

% create the lattice
real_matrix=zeros(matrix_size+1, matrix_size+1);
pred_matrix=zeros(matrix_size+1, matrix_size+1);

real_points = [test_real_y(:, 1), test_real_y(:, 3)];
pred_points = [test_pred_y(:, 1), test_pred_y(:, 3)];

% limit real_points and pred_points in the range of [min_value, max_value]
real_points = max(min(real_points, max_value), min_value);
pred_points = max(min(pred_points, max_value), min_value);

% count the number of points in each box
for pt = 1:length(real_points)
    real_x = floor(abs((real_points(pt, 1)-lower_) / dv_dt));
    real_y = floor(abs((real_points(pt, 2)-lower_) / dv_dt));
    pred_x = floor(abs((pred_points(pt, 1)-lower_) / dv_dt));
    pred_y = floor(abs((pred_points(pt, 2)-lower_) / dv_dt));

    if real_x == 0
        real_x = 1;
    end
    if real_y == 0
        real_y = 1;
    end
    if pred_x == 0
        pred_x = 1;
    end
    if pred_y == 0
        pred_y = 1;
    end

    real_matrix(real_x, real_y) = real_matrix(real_x, real_y)+1;
    pred_matrix(pred_x, pred_y) = pred_matrix(pred_x, pred_y)+1;
end

real_matrix = (reshape(real_matrix, 1, []))./testing_length;
pred_matrix = (reshape(pred_matrix, 1, []))./testing_length;

DV = sum(sqrt((real_matrix-pred_matrix).^2));


figure();

plot(real_points(:,1), real_points(:,2), 'r-', LineWidth=1);
hold on
plot(pred_points(:,1), pred_points(:, 2), 'b--', LineWidth=1);

xlim([min_value-dv_dt, max_value+dv_dt]);
ylim([min_value-dv_dt, max_value+dv_dt]);
for iii = 1:120
    xline(min_value+dv_dt*iii);
    yline(min_value+dv_dt*iii);
end

legend('real', 'prediction')
title('lorenz system long term prediction')
text(2, -2,  ['DV=', num2str(DV)]);

File: python/reservoir.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  2 20:23:17 2024

@author: zmzhai
"""

import pandas as pd
import numpy as np
import os
import random
from sklearn.preprocessing import StandardScaler, MinMaxScaler
import matplotlib.pyplot as plt
import pickle
import networkx as nx
import utils

# dimension of the system
dim = 3

# reservoir hyperparameters
pkl_file = open('./save_opt/rc_opt_chaos_lorenz.pkl', 'rb') 
opt_results = pickle.load(pkl_file)
pkl_file.close()
opt_params = opt_results['params']

n = 200
eig_rho = opt_params['eig_rho']
gamma = opt_params['gamma']
alpha = opt_params['alpha']
beta = 10 ** opt_params['beta']
d = opt_params['d']
noise_a = 10 ** opt_params['noise_a']

# data simulation
train_length = 50000
test_length = 10000
short_prediction_length = 700

dt = 0.01
t_end = 1000
t_all = np.arange(0, t_end, dt)
x0 = [(0.8 - 0.5) * np.random.rand() + 0.5, 0.3 * np.random.rand(), (10 - 8) * np.random.rand() + 8]

ts = utils.rk4(utils.func_lorenz, x0, t_all, params=np.array([]))
# discard the transient
ts = ts[5000:, :]
# normalization
standard_scaler = StandardScaler()
ts_train = standard_scaler.fit_transform(ts)

# ts_train = ts

# fig, ax = plt.subplots(3, 1, constrained_layout=True)
# plot_length = 3000
# ax[0].plot(ts_train[:plot_length, 0])
# ax[1].plot(ts_train[:plot_length, 1])
# ax[2].plot(ts_train[:plot_length, 2])
# ax[0].set_ylabel('x')
# ax[1].set_ylabel('y')
# ax[2].set_ylabel('z')
# ax[2].set_xlabel('step')

# reservoir computer configuration
Win = np.random.uniform(-gamma, gamma, (n, dim))
graph = nx.erdos_renyi_graph(n, d, 42, False)
for (u, v) in graph.edges():
    graph.edges[u, v]['weight'] = np.random.normal(0.0, 1.0)
A = nx.adjacency_matrix(graph).todense()
rho = max(np.linalg.eig(A)[0])
A = (eig_rho / abs(rho)) * A

# train
r_train = np.zeros((n, train_length))
y_train = np.zeros((dim, train_length))
r_end = np.zeros((n, 1))

train_x = np.zeros((train_length, dim))
train_y = np.zeros((train_length, dim))

train_y[:, :] = ts_train[1:train_length+1, :]

noise = noise_a * np.random.randn(*ts_train[:train_length, :].shape)
# Adding the noise to the ts_train data
ts_train[:train_length, :] += noise

train_x[:, :] = ts_train[:train_length, :]

train_x = np.transpose(train_x)
train_y = np.transpose(train_y)

r_all = np.zeros((n, train_length + 1))

for ti in range(train_length):
    r_all[:, ti+1] = (1 - alpha) * r_all[:, ti] + \
        alpha * np.tanh( np.dot(A, r_all[:, ti]) + np.dot(Win, train_x[:, ti])  )

r_out = r_all[:, 1:]
r_end[:] = r_all[:, -1].reshape(-1, 1)

r_train[:, :] = r_out
y_train[:, :] = train_y[:, :]

Wout = np.dot(np.dot(y_train, np.transpose(r_train)), np.linalg.inv(np.dot(r_train, np.transpose(r_train)) + beta * np.eye(n)) )

# train_preditions = np.dot(Wout, r_out)

# fig, ax = plt.subplots(1, 1, constrained_layout=True)
# ax.plot(train_preditions[0, 20000:25000])
# ax.plot(train_y[0, 20000:25000])

# test

testing_start = train_length + 1

test_pred = np.zeros((test_length, dim))
test_real = np.zeros((test_length, dim))

test_real[:, :] = ts_train[testing_start:testing_start+np.shape(test_real)[0], :]

r = r_end

u = np.zeros((dim, 1))
u[:] = ts_train[train_length, :].reshape(-1, 1)
for ti in range(test_length-1):
    r = (1 - alpha) * r + alpha * np.tanh(np.dot(A, r) + np.dot(Win, u))
    
    pred = np.dot(Wout, r)
    test_pred[ti, :] = pred.reshape(dim, -1).ravel()
    
    u[:] = pred

rmse = utils.rmse_calculation(test_pred[:short_prediction_length,:], test_real[:short_prediction_length,:])

# plot short term prediction
fig, ax = plt.subplots(3, 1, constrained_layout=True)

ax[0].plot(test_real[:short_prediction_length,0], label='real')
ax[0].plot(test_pred[:short_prediction_length,0], label='pred')
ax[0].set_ylabel('x')
ax[0].legend()

ax[1].plot(test_real[:short_prediction_length,1], label='real')
ax[1].plot(test_pred[:short_prediction_length,1], label='pred')
ax[1].set_ylabel('y')

ax[2].plot(test_real[:short_prediction_length,2], label='real')
ax[2].plot(test_pred[:short_prediction_length,2], label='pred')
ax[2].set_ylabel('z')
ax[2].set_xlabel('step')





















